### DISCUSSION TITLE
Failed to build wheel for bcrypt - can't install docker-compose

### DISCUSSION THREAD

**Discussion by swizzlest** [teslamate-org/teslamate#2881] [Getting started] [answered]
Fresh install, latest version.  I get to 'sudo pip3 -v install docker-compose' on the linked website for setting up docker on raspberry pi (https://dev.to/rohansawant/installing-docker-and-docker-compose-on-the-raspberry-pi-in-5-simple-steps-3mgl) and get this output:

pi@raspberrypi:~ $ sudo pip3 install docker-compose
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting docker-compose
  Using cached https://www.piwheels.org/simple/docker-compose/docker_compose-1.29.2-py2.py3-none-any.whl (114 kB)
Collecting jsonschema<4,>=2.5.1
  Using cached https://www.piwheels.org/simple/jsonschema/jsonschema-3.2.0-py2.py3-none-any.whl (56 kB)
Collecting docopt<1,>=0.6.1
  Using cached https://www.piwheels.org/simple/docopt/docopt-0.6.2-py2.py3-none-any.whl (13 kB)
Collecting docker[ssh]>=5
  Using cached https://www.piwheels.org/simple/docker/docker-6.0.0-py3-none-any.whl (147 kB)
Requirement already satisfied: distro<2,>=1.5.0 in /usr/lib/python3/dist-packages (from docker-compose) (1.5.0)
Collecting python-dotenv<1,>=0.13.0
  Using cached python_dotenv-0.21.0-py3-none-any.whl (18 kB)
Requirement already satisfied: requests<3,>=2.20.0 in /usr/lib/python3/dist-packages (from docker-compose) (2.25.1)
Collecting websocket-client<1,>=0.32.0
  Using cached https://www.piwheels.org/simple/websocket-client/websocket_client-0.59.0-py2.py3-none-any.whl (67 kB)
Collecting dockerpty<1,>=0.4.1
  Using cached https://www.piwheels.org/simple/dockerpty/dockerpty-0.4.1-py3-none-any.whl (16 kB)
Collecting texttable<2,>=0.9.0
  Using cached https://www.piwheels.org/simple/texttable/texttable-1.6.4-py2.py3-none-any.whl (10 kB)
Collecting PyYAML<6,>=3.10
  Using cached https://www.piwheels.org/simple/pyyaml/PyYAML-5.4.1-cp39-cp39-linux_armv7l.whl (45 kB)
Collecting requests<3,>=2.20.0
  Using cached https://www.piwheels.org/simple/requests/requests-2.28.1-py3-none-any.whl (62 kB)
Collecting packaging>=14.0
  Using cached https://www.piwheels.org/simple/packaging/packaging-21.3-py3-none-any.whl (40 kB)
Requirement already satisfied: urllib3>=1.26.0 in /usr/lib/python3/dist-packages (from docker[ssh]>=5->docker-compose) (1.26.5)
Collecting paramiko>=2.4.3
  Using cached https://www.piwheels.org/simple/paramiko/paramiko-2.11.0-py2.py3-none-any.whl (215 kB)
Requirement already satisfied: six>=1.3.0 in /usr/lib/python3/dist-packages (from dockerpty<1,>=0.4.1->docker-compose) (1.16.0)
Collecting pyrsistent>=0.14.0
  Using cached https://www.piwheels.org/simple/pyrsistent/pyrsistent-0.18.1-cp39-cp39-linux_armv7l.whl (101 kB)
Requirement already satisfied: setuptools in /usr/lib/python3/dist-packages (from jsonschema<4,>=2.5.1->docker-compose) (52.0.0)
Collecting attrs>=17.4.0
  Using cached https://www.piwheels.org/simple/attrs/attrs-22.1.0-py2.py3-none-any.whl (58 kB)
Collecting pyparsing!=3.0.5,>=2.0.2
  Using cached https://www.piwheels.org/simple/pyparsing/pyparsing-3.0.9-py3-none-any.whl (98 kB)
Collecting pynacl>=1.0.1
  Using cached https://www.piwheels.org/simple/pynacl/PyNaCl-1.5.0-cp39-cp39-linux_armv7l.whl (585 kB)
Requirement already satisfied: cryptography>=2.5 in /usr/lib/python3/dist-packages (from paramiko>=2.4.3->docker[ssh]>=5->docker-compose) (3.3.2)
Collecting bcrypt>=3.1.3
  Using cached bcrypt-4.0.0.tar.gz (25 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Collecting cffi>=1.4.1
  Using cached https://www.piwheels.org/simple/cffi/cffi-1.15.1-cp39-cp39-linux_armv7l.whl (364 kB)
Collecting pycparser
  Using cached https://www.piwheels.org/simple/pycparser/pycparser-2.21-py2.py3-none-any.whl (119 kB)
Collecting charset-normalizer<3,>=2
  Using cached https://www.piwheels.org/simple/charset-normalizer/charset_normalizer-2.1.1-py3-none-any.whl (39 kB)
Requirement already satisfied: certifi>=2017.4.17 in /usr/lib/python3/dist-packages (from requests<3,>=2.20.0->docker-compose) (2020.6.20)
Requirement already satisfied: idna<4,>=2.5 in /usr/lib/python3/dist-packages (from requests<3,>=2.20.0->docker-compose) (2.10)
Building wheels for collected packages: bcrypt
  Building wheel for bcrypt (PEP 517) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3 /tmp/tmp92f6fdta_in_process.py build_wheel /tmp/tmp2ym1irq6
       cwd: /tmp/pip-install-yn0v49mv/bcrypt_e4b7a20b595d4e6c91b63cae54a0dc69
  Complete output (58 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-armv7l-cpython-39
  creating build/lib.linux-armv7l-cpython-39/bcrypt
  copying src/bcrypt/__about__.py -> build/lib.linux-armv7l-cpython-39/bcrypt
  copying src/bcrypt/__init__.py -> build/lib.linux-armv7l-cpython-39/bcrypt
  running egg_info
  writing src/bcrypt.egg-info/PKG-INFO
  writing dependency_links to src/bcrypt.egg-info/dependency_links.txt
  writing requirements to src/bcrypt.egg-info/requires.txt
  writing top-level names to src/bcrypt.egg-info/top_level.txt
  reading manifest file 'src/bcrypt.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  warning: no previously-included files found matching 'requirements.txt'
  warning: no previously-included files found matching 'release.py'
  warning: no previously-included files found matching 'mypy.ini'
  warning: no previously-included files matching '*' found under directory '.github'
  warning: no previously-included files matching '*' found under directory '.circleci'
  warning: no previously-included files found matching 'src/_bcrypt/target'
  warning: no previously-included files matching '*' found under directory 'src/_bcrypt/target'
  adding license file 'LICENSE'
  writing manifest file 'src/bcrypt.egg-info/SOURCES.txt'
  copying src/bcrypt/_bcrypt.pyi -> build/lib.linux-armv7l-cpython-39/bcrypt
  copying src/bcrypt/py.typed -> build/lib.linux-armv7l-cpython-39/bcrypt
  running build_ext
  running build_rust

      =============================DEBUG ASSISTANCE=============================
      If you are seeing a compilation error please try the following steps to
      successfully install bcrypt:
      1) Upgrade to the latest pip and try again. This will fix errors for most
         users. See: https://pip.pypa.io/en/stable/installing/#upgrading-pip
      2) Ensure you have a recent Rust toolchain installed. bcrypt requires
         rustc >= 1.56.0.

      Python: 3.9.2
      platform: Linux-5.15.32-v7l+-armv7l-with-glibc2.31
      pip: n/a
      setuptools: 65.3.0
      setuptools_rust: 1.5.1
      rustc: n/a
      =============================DEBUG ASSISTANCE=============================

  error: can't find Rust compiler

  If you are using an outdated pip version, it is possible a prebuilt wheel is available for this package but pip is not able to install from it. Installing from the wheel would avoid the need for a Rust compiler.

  To update pip, run:

      pip install --upgrade pip

  and then retry package installation.

  If you did intend to build this package from source, try installing a Rust compiler from your system package manager and ensure it is on the PATH during installation. Alternatively, rustup (available at https://rustup.rs) is the recommended way to download and update the Rust compiler toolchain.

  This package requires Rust >=1.56.0.
  ----------------------------------------
  ERROR: Failed building wheel for bcrypt
Failed to build bcrypt
ERROR: Could not build wheels for bcrypt which use PEP 517 and cannot be installed directly

I have tried upgrading pip.  I have tried starting over.  I have tried 64bit and 32bit images.  I have tried installing docker-compose v2.  Nothing seems to rectify.  Anyone have any ideas?

Thank you!

**Comment by swizzlest** [comment, marked as answer]
In case anyone else runs into this, I have no idea what I did wrong, but apparently it is something as I'm the only thread like this.  I did the steps 4-5 times and had this result every time.

I eventually installed docker-compose with apt using

sudo apt install docker-compose

Then it all worked fine.

**Reply by aaronrothschild-pd** [reply]
Ugh, so easy. I reflashed the pi, tried rust, etc....nothing worked.  Thank you!

**Reply by geoscoutCJ** [reply]
Thank you!! Working on a totally different project but was getting basically the same thing. Head banged against many walls....😄

**Reply by russmcg** [reply]
This worked, thank you!

**Reply by hiddit** [reply]
Wow, this helped me out so much. Also for a other project with the same issue. Thank you!!

**Reply by MasterMik** [reply]
Nice, it worked.

**Reply by chefdog** [reply]
Thank you!

**Reply by josaloma** [reply]
Thanks. Worked!

**Reply by stanifert** [reply]
This worked for me when nothing else was. Thanks!

**Comment by AmedeeBulle** [comment]
For the ones who still want to install with pip to get the latest 1.x release... 
The problem comes with bcrypt 4.0.0 for which there is no wheel available. You can pre-install the latest 3.x version of bcrypt before installing compose:
```
sudo pip3 install -U "bcrypt<4.0.0"
```

**Reply by dirkvdplas** [reply]
Yes, that worked like a charm!

**Reply by giancake** [reply]
it worked on my pi I fixed the error

**Reply by Arcangel96** [reply]
Man ere un máster ❤️❤️❤️❤️❤️

**Reply by taupiknr26** [reply]
it worked, you saves my day, thanks
